helix-core/src/main/java/org/apache/helix/controller/dataproviders/BaseControllerDataProvider.java

```diff
@@ -82,7 +84,7 @@
   private ExecutorService _asyncTasksThreadPool;
 
   // A map recording what data has changed
-  protected Map<HelixConstants.ChangeType, Boolean> _propertyDataChangedMap;
+  protected Map<HelixConstants.ChangeType, AtomicBoolean> _propertyDataChangedMap;
```

Comment:
I think the atomicBoolean here is not essential. To fix this bug, we don't really need atomicBoolean. It is essential to use a concurrentHashMap as notifyDataChange() will update value in this map to true in an async way. Value changed in ConcurrentHashMap is considered as safe publication anyway.

Reply:
That is not essential. Just to make the code simpler and preventing some of the potential bugs.
Just to confirm, are you suggesting that we convert back to Boolean?

Reply:
@jiajunwang I know it, the context is not obviously known to everyone. @kaisun2000 The real reason is for the sake of performance, we want to a lock-free condition where the `notifyDataChange` runs faster (it simply sets a flag without acquiring lock). In the other hand, the checker (waits for the condition) needs to have the update-to-date information in the multi-threading world. AtomicBoolean is the recommended way to avoid lock and achieve the goal. Some similar discussions are talked about here https://stackoverflow.com/questions/32935548/condition-variable-without-mutex-in-a-lock-free-implementation